bindings/matrix-sdk-ffi/src/client_builder.rs

```diff
@@ -238,6 +378,109 @@ impl ClientBuilder {
     pub async fn build(self: Arc<Self>) -> Result<Arc<Client>, ClientBuildError> {
         Ok(Arc::new(self.build_inner().await?))
     }
+
+    /// Finish the building of the client and attempt to log in using the
+    /// provided [`QrCodeData`].
+    ///
+    /// This method will build the client and immediately attempt to log the
+    /// client in using the provided [`QrCodeData`] using the login
+    /// mechanism described in [MSC4108]. As such this methods requires OIDC
+    /// support as well as sliding sync support.
+    ///
+    /// The usage of the progress_listener is required to transfer the
+    /// [`CheckCode`] to the existing client.
+    ///
+    /// [MSC4108]: https://github.com/matrix-org/matrix-spec-proposals/pull/4108
+    pub async fn build_with_qr_code(
+        self: Arc<Self>,
+        qr_code_data: &QrCodeData,
+        oidc_configuration: &OidcConfiguration,
+        progress_listener: Box<dyn QrLoginProgressListener>,
+    ) -> Result<Arc<Client>, HumanQrLoginError> {
+        if let QrCodeModeData::Reciprocate { homeserver_url } = &qr_code_data.inner.mode_data {
```

Comment:
style proposal to spare one full level of indent:

 `let QrCodeModeData::Reciprocate { homeserver_url } = &qr_code_data.inner.mode_data else { return Err(HumanQrLoginError::InvalidQrCode); }`

Reply:
I actually prefer it this way, as long as it's only one indent.